src/utils/metrics/metrics.go

```diff
@@ -20,113 +20,289 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-// Package metrics [everything related to metrics goes here]
+// Package metrics collects and reports job metrics.
 package metrics
 
 import (
-	"context"
+	"fmt"
+	"io"
+	"sort"
 	"sync"
-	"time"
+	"text/tabwriter"
 )
 
-// supported default metrics
+// Stat is the type of statistical metrics.
+type Stat int
+
 const (
-	Traffic          = "traffic"
-	ProcessedTraffic = "processed_traffic"
+	RequestsAttemptedStat Stat = iota
+	RequestsSentStat
+	ResponsesReceivedStat
+	BytesSentStat
+
+	NumStats
 )
 
-// Storage is a general struct to store custom metrics
-type Storage struct {
-	trackers map[string]*metricTracker // map by metric type
+// String representation of the Stat.
+func (s Stat) String() string {
+	if !s.valid() {
+		return "<unknown>"
+	}
+
+	return [...]string{
+		"requests attempted",
+		"requests sent",
+		"responses received",
+		"bytes sent",
+	}[s]
+}
+
+func (s Stat) valid() bool { return s >= RequestsAttemptedStat && s < NumStats }
+
+type dimensions struct {
+	jobID  string
+	target string
+}
+
+// Reporter gathers metrics across jobs and reports them.
+// Concurrency-safe.
+type Reporter struct {
+	metrics [NumStats]sync.Map // Array of metrics by Stat. Each metric is a map of uint64 values by dimensions.
+
+	clientID string
+}
+
+// NewReporter creates a new Reporter.
+func NewReporter(clientID string) *Reporter { return &Reporter{clientID: clientID} }
+
+// Sum returns a total sum of metric s.
+func (r *Reporter) Sum(s Stat) uint64 {
+	if r == nil || !s.valid() {
```

Comment:
What is the use case for these checks? We don't have any user-defined data coming into this API so it would be better to panic here to catch issues faster and have a clear branch prediction pattern for the compiler that should allow for better optimizations

Reply:
I initially assumed you want this check for a case of disabling metrics for encrypted jobs but I see that you still pass a correct accumulator handle there so it shouldn't be needed

Reply:
same obviously goes for all other checks of this type

Reply:
Yes, the intention was to disable collection for encrypted jobs. I must have forgotten about this detail, let me cross-check.

As for panicking I also tend to agree, guess I'll just drop the validations.